unpack-trees.c

```diff
@@ -2512,6 +2519,55 @@ static void show_stage_entry(FILE *o,
 }
 #endif
 
+static int merge_sparse_dirs(const struct cache_entry *base,
+		      const struct cache_entry *left,
+		      const struct cache_entry *right,
+		      struct unpack_trees_options *o)
+{
+	const struct object_id *empty_tree_oid;
+	struct merge_options opt;
+	struct cache_entry *merge;
+	struct tree *base_tree, *left_tree, *right_tree;
+	struct merge_result result;
+	struct strbuf conflict_path = STRBUF_INIT;
+	int ret;
+
+	assert(left != NULL);
+
+	init_merge_options(&opt, o->src_index->repo);
+	opt.ancestor = "base";
+	opt.branch1 = "left";
+	opt.branch2 = "right";
+	memset(&result, 0, sizeof(result));
+
+	empty_tree_oid = o->src_index->repo->hash_algo->empty_tree;
+	base_tree = lookup_tree(o->src_index->repo, base ? &base->oid : empty_tree_oid);
+	left_tree = lookup_tree(o->src_index->repo, &left->oid);
+	right_tree = lookup_tree(o->src_index->repo, right ? &right->oid : empty_tree_oid);
+
+	merge_incore_nonrecursive(&opt, base_tree, left_tree, right_tree, &result);
```

Comment:
@derrickstolee After thinking about this more, `merge-ort` isn't _quite_ right for this case (as it is now, way) - `merge-ort` performs content merges, whereas `twoway_merge` does not and has its own set of rules for how to "merge" files. Since `stash` only uses the `oneway_merge` case of `read-tree`, I'd like to revert back to d25d0f2e1020e97eea8c51acfef86b9286392113 (with the code style cleanup you've pointed out). That should still get the performance benefit in `stash` (I'll verify that with the tests, though), as well as not introducing functionality that creates mismatches between sparse index and non-sparse index checkouts.

Reply:
That works for me. Let's not make any unnecessary changes here.